Update dataset file locations with Wasabi URLs

pli888 opened the pull request:
# Pull request for issue: #1279 and #1280

This is a pull request for the following functionalities:

* A command-line tool in `files-metadata-console` that will update Dataset `ftp_site` and File `location` in the PostgreSQL database with Wasabi URLs

## How to test?

### Dev environment

#### Manual testing

From your `gigadb-website` project repository, check out the code from this PR into a new branch:
```
$ git checkout -b pli888-wasabi-file-urls develop
$ git pull https://github.com/pli888/gigadb-website.git wasabi-file-urls
```

Deploy your GigaDB `dev` environment:
```
$ ./up.sh
```

Go to `files-metadata-console` directory and create it's configuration:
```
$ cd gigadb/app/tools/files-metadata-console
# Create .env file
$ ./configure
# Add your Gitlab project name and private token into .env file
# Execute configure again to a .secrets file
$ ./configure
# Install tool dependencies
$ docker-compose run --rm composer composer install
```
> You should now have a .secrets file created in `files-metadata-console`.

Let's check that datasets have files which are stored in CNGB. Go to http://gigadb.gigasciencejournal.com:9170/dataset/100006 in your browser and click on `Files` tab. If you click on the `readme.txt` link then the file will be downloaded from CNGB using this link:
https://ftp.cngb.org/pub/gigadb/pub/10.5524/100001_101000/100006/readme.txt.

Run the tool in file-metadata-console in `apply` mode to update dataset file URLs so they can be fetched from Wasabi storage. The command line parameters will tell the tool to start processing in batches of 3 datasets, and exclude datasets 100020 and 100039:
```
# Run tool to update URLs
$ docker-compose run --rm files-metadata-console ./yii update/urls --prefix=https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live --separator=/pub/ --next=3 --exclude='100020,100039' --apply
Creating files-metadata-console_files-metadata-console_run ... done
        Transforming ftp_site for dataset 100006... 
DONE

        Transforming file locations for dataset 100006...
DONE (7/7)

        Transforming ftp_site for dataset 100056... 
DONE

        Transforming file locations for dataset 100056...
DONE (5/5)

        Transforming ftp_site for dataset 100094... 
DONE

        Transforming file locations for dataset 100094...
DONE (5/5)
```

Since datasets 100020 and 100039 have been excluded, the next dataset to be processed after 100006 is 100056 which will have all its file URLs updated. The final updated dataset is 100094 since this is the third dataset in the batch of 3 to be processed.

Check dataset file URLs have been updated by going to http://gigadb.gigasciencejournal.com:9170/dataset/100056 in your browser and click on `Files` tab. If you right-click on the `readme.txt` link and open it in a new browser tab then you should see the contents of the readme file fetched from https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live/pub/10.5524/100001_101000/100056/readme.txt.

Re-run same command without excluding DOI and increasing the batch number to 4 but run in dry mode by not using the `apply` flag:
```
$ docker-compose run --rm files-metadata-console ./yii update/urls --prefix=https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live --separator=/pub/ --next=4
Creating files-metadata-console_files-metadata-console_run ... done
Running tool in dry run mode
        Transforming ftp_site for dataset 100020... 
DONE

        Transforming file locations for dataset 100020...
DONE (7/7)

        Transforming ftp_site for dataset 100039... 
DONE

        Transforming file locations for dataset 100039...
DONE (24/24)

        Transforming ftp_site for dataset 100142... 
DONE

        Transforming file locations for dataset 100142...
DONE (4/4)

        Transforming ftp_site for dataset 100245... 
DONE

        Transforming file locations for dataset 100245...
DONE (17/17)

```

This time, the tool goes back to process datasets 100020 and 100039, and also processes datasets 100142 and 100245. But because dry run mode is used then no URL changes have been made. Take a look at the readme file for http://gigadb.gigasciencejournal.com:9170/dataset/100020 to confirm this.

Re-running the same command with `apply` flag will make the relevant changes to the Dataset and File database tables:
```
$ docker-compose run --rm files-metadata-console ./yii update/urls --prefix=https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live --separator=/pub/ --next=4 --apply
Creating files-metadata-console_files-metadata-console_run ... done
        Transforming ftp_site for dataset 100020... 
DONE

        Transforming file locations for dataset 100020...
DONE (7/7)

        Transforming ftp_site for dataset 100039... 
DONE

        Transforming file locations for dataset 100039...
DONE (24/24)

        Transforming ftp_site for dataset 100142... 
DONE

        Transforming file locations for dataset 100142...
DONE (4/4)

        Transforming ftp_site for dataset 100245... 
DONE

        Transforming file locations for dataset 100245...
DONE (17/17)

```

In a browser, go to http://gigadb.gigasciencejournal.com:9170/dataset/100245 and click on `Files` tab. If you click a file such as IMG_2767.jpg to open it in a new web browser tab, you should see that it is fetched via a Wasabi URL.

#### Unit and functional tests

Before we can run the tests, we need to build Codeception actor classes:
```
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept build
Creating files-metadata-console_files-metadata-console_run ... done
Building Actor classes for suites: functional, unit
 -> FunctionalTesterActions.php generated successfully. 254 methods added
\FunctionalTester includes modules: Asserts, Filesystem, Yii2, Cli, Db
```

Execute all unit tests:
```
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/unit/DatasetFilesURLUpdaterTest.php
```

Execute all functional tests:
```
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/functional/ReplaceFileUrlSubstringWithPrefixCest.php
```

### Staging environment

Instantiate the AWS resources for your staging environment:
```
$ cd /path/to/gigadb-website/ops/infrastructure/envs/staging

# Copy terraform files to staging environment
$ ../../../scripts/tf_init.sh --project gigascience/forks/your-gigadb-website --env staging

You need to specify the path to the ssh private key to use to connect to the EC2 instance:
~/.ssh/your-id-rsa-aws.pem

You need to specify your GitLab username:
your-repo-name

You need to specify an AWS region:
ap-northeast-1

# Provision with Terraform
$ terraform plan
$ terraform apply
$ terraform refresh

# Copy ansible files
$ ../../../scripts/ansible_init.sh --env staging

# Provision bastion ec2 server and restore RDS with latest database backup
$ ansible-playbook -i ../../inventories bastion_playbook.yml

# Provision webapp ec2 server with ansible
$ TF_KEY_NAME=private_ip ansible-playbook -i ../../inventories webapp_playbook.yml
```

Run Gitlab CI/CD pipeline for your `pli888-wasabi-file-urls` branch.

When `sd_gigadb` and `FileMetadataConsoleDeployStaging` have completed in your pipeline, check the website for your staging server. You will see that every dataset page, e.g. `/dataset/102401` will have file URLs beginning with `https://ftp.cngb.org/pub/gigadb/pub/10.5524`.

SSH into your bastion server:
```
$ ssh -i "~/.ssh/your-id-rsa-aws.pem" centos@ec2_bastion_public_ip
```

Due to memory issues, we need to drop database triggers otherwise the tool will hang during dataset processing

```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists file_finder_trigger on file RESTRICT'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists sample_finder_trigger on sample RESTRICT'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists dataset_finder_trigger on dataset RESTRICT'
```

Before file location URL transformation, take a look at /dataset/100001 in your staging server and its readme_100001.txt file. You should see that it is located by its https://ftp.cngb.org link.

Now start transforming the URLs for file locations:
```
$ docker run --rm "registry.gitlab.com/$GITLAB_PROJECT/production-files-metadata-console:latest" ./yii update/urls --prefix=https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live --separator=/pub/ --exclude='100375,100396,100446,100584,100747,100957,102311' --stop=200002 --next=30 --apply
```
>A number of datasets have curation errors and their DOIs are excluded from processing until curators have fixed these problems. When the tool reaches dataset DOI 200002 then it needs to stop execution as datasets 200002 and above are not "real" datasets and should not be transformed.

If you go to /dataset/100001 and now look at its readme_100001.txt, you should see it resides at https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live/pub/10.5524/100001_101000/100001/readme_100001.txt.

We need to manually update file location URLs for dataset 100396 since it contains nearly 200,000 files which the tool is not able to handle:
```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c "Update file set location = REPLACE(location, 'https://ftp.cngb.org/pub/gigadb/pub/', 'https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live/pub/') where dataset_id = 629;"
UPDATE 197214
```

When you are done with processing then the database triggers need to be re-created:
```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'create trigger file_finder_trigger after insert or update or delete or truncate on file for each statement execute procedure refresh_file_finder()'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'create trigger sample_finder_trigger after insert or update or delete or truncate on sample for each statement execute procedure refresh_sample_finder()'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'create trigger dataset_finder_trigger after insert or update or delete or truncate on dataset for each statement execute procedure refresh_dataset_finder()'
```

### Live environment

Instantiate the AWS resources for your live environment:
```
$ cd /path/to/gigadb-website/ops/infrastructure/envs/live

# Copy terraform files to staging environment
$ ../../../scripts/tf_init.sh --project gigascience/forks/your-gigadb-website --env live

You need to specify the path to the ssh private key to use to connect to the EC2 instance:
~/.ssh/your-id-rsa-aws.pem

You need to specify your GitLab username:
your-repo-name

You need to specify an AWS region:
ap-northeast-1

# Provision with Terraform
$ terraform plan
$ terraform apply
$ terraform refresh

# Copy ansible files
$ ../../../scripts/ansible_init.sh --env live

# Provision bastion ec2 server and restore RDS with latest database backup
$ ansible-playbook -i ../../inventories bastion_playbook.yml

# Provision webapp ec2 server with ansible
$ TF_KEY_NAME=private_ip ansible-playbook -i ../../inventories webapp_playbook.yml
```

Run Gitlab CI/CD pipeline for your `pli888-wasabi-file-urls` branch.

When `sd_gigadb` and `FileMetadataConsoleDeployStaging` have completed in your pipeline, check the website for your staging server. You will see that every dataset page, e.g. `/dataset/102401` will have file URLs beginning with `https://ftp.cngb.org/pub/gigadb/pub/10.5524`.

SSH into your bastion server:
```
$ ssh -i "~/.ssh/your-id-rsa-aws.pem" ssh centos@ec2_bastion_public_ip
```

Due to memory issues, we need to drop database triggers otherwise the tool will hang during dataset processing

```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists file_finder_trigger on file RESTRICT'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists sample_finder_trigger on sample RESTRICT'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'drop trigger if exists dataset_finder_trigger on dataset RESTRICT'
```

Now start transforming the URLs for file locations:
```
$ docker run --rm "registry.gitlab.com/$GITLAB_PROJECT/production-files-metadata-console:latest" ./yii update/urls --prefix=https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live --separator=/pub/ --exclude='100375,100396,100446,100584,100747,100957,102311' --stop=200002 --next=30 --apply
```
>A number of datasets have curation errors and their DOIs are excluded from processing until curators have fixed these problems. When the tool reaches dataset DOI 200002 then it needs to stop execution as datasets 200002 and above are not "real" datasets and should not be transformed. 

We need to manually update file location URLs for dataset 100396 since it contains nearly 200,000 files which the tool is not able to handle:
```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c "Update file set location = REPLACE(location, 'https://ftp.cngb.org/pub/gigadb/pub/', 'https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live/pub/') where dataset_id = 629;"
UPDATE 197214
```

Also, need to update `ftp_site` field in `dataset` table for dataset 100396:
```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c "Update dataset set ftp_site = REPLACE(ftp_site, 'https://ftp.cngb.org/pub/gigadb/pub/', 'https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/live/pub/') where identifier = '100396';"
UPDATE 1
```

When you are done with processing then the database triggers need to be re-created:
```
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'create trigger file_finder_trigger after insert or update or delete or truncate on file for each statement execute procedure refresh_file_finder()'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'create trigger sample_finder_trigger after insert or update or delete or truncate on sample for each statement execute procedure refresh_sample_finder()'
$ docker run --rm  --env-file ./db-env registry.gitlab.com/$GITLAB_PROJECT/production_pgclient:$GIGADB_ENV -c 'create trigger dataset_finder_trigger after insert or update or delete or truncate on dataset for each statement execute procedure refresh_dataset_finder()'
```

Don't forget to get rid of AWS resources when testing is finished:

```
# Terminate AWS resources afterwards
$ terraform destroy

```

## How have functionalities been implemented?

How Dataset ftp_site and File location attributes are transformed is implemented in the `DatasetFilesURLUpdater` component class. There is a `build` factory function to create this component class and configures it to perform its URL transformations in dry-run mode or not.

The function `getNextPendingDatasets()` will identify datasets with files that have old CNGB location URLs, returning these dataset DOIs. The functions `replaceFileLocationsForDataset()` and `replaceFTPSiteForDataset()` will then transform Dataset ftp_site and File location attributes.

The UpdateController class has a new function called `actionUrls()` which provides the command line interface for users to transform File location URLs.

## Any issues with implementation?

Dataset 100396 contains nearly 200,000 files which is too many file objects for the tool to handle, causing it to hang when it reaches this dataset DOI. To resolve this problem, an UPDATE SQL command is provided to update the File location URLs for this dataset.

The RDS terraform script had to be updated to use PostgreSQL 11.16 since the version 11.13 seems to be not available anymore.

## Any changes to automated tests?

New unit and functional tests have been written to check the functionality of the new Component class and new Controller function.

All tests in `file-metadata-console` should now be using the gigadb_testdata database.

## Any changes to documentation?

The README file contains information on how to use the tool.

pli888 committed:
Update file URL with prefix

pli888 committed:
Save changes made to URLs

pli888 committed:
Update implementation for updating file locations

pli888 committed:
Check mandatory options are provided in command

pli888 committed:
Add function to get pending datasets

pli888 committed:
Query two or more dois

pli888 committed:
Add where clauses to look for other ftp URLs

pli888 committed:
Return array of dois in getNextPendingDatasets function

pli888 committed:
Add batch functionality

pli888 committed:
Save new ftp_site attribute into dataset

pli888 committed:
Add a dry run mode using apply command line flag

pli888 committed:
Add ability to exclude datasets from having file URLs updated

pli888 committed:
Move getNextPendingDatasets into a unit test

pli888 committed:
Add another unit test function

pli888 committed:
Add new unit test function

This test function checks updating of file URL with a prefix.

pli888 committed:
Update functional test

pli888 committed:
Use pg_restore to restore gigadb_testdata to original state

pli888 committed:
Add Db configuration for codeception

pli888 committed:
WIP improve functional test

pli888 committed:
Update configuration for unit and func tests

pli888 committed:
Finish functional test when tool is executed twice

pli888 committed:
Provide excluded DOIs cmd-line parameter value as string

Excluded DOIs should be defined on command line as a string with
commas separating the DOIs. This string will be split into an
array of DOIs.

pli888 committed:
Add dry run functional test

pli888 committed:
Put file URL updating code into a new component class

pli888 committed:
Move update ftp_site code into own function

pli888 committed:
Created functions for updating ftp_site and file location

pli888 committed:
Improve exception handling

pli888 committed:
Provide bucket name and directories in a constant

pli888 committed:
Throw exception when dataset has unexpected ftp_site

pli888 committed:
Provide new implementation for replaceFileLocationPrefix function

pli888 committed:
Update function to replace file locations

pli888 committed:
Create unit test for updating dataset ftp_site

pli888 committed:
Tidy up comments in code

pli888 committed:
Improve logging to stdout

pli888 committed:
Rename function to replaceFileLocationsForDataset

pli888 committed:
Update cmd-line parameter documentation

pli888 committed:
Update dbname to gigadb_testdata database

pli888 committed:
Update dependencies listed in composer

pli888 committed:
Change dataset ftp_site and file location

pli888 committed:
Fix functional tests in files-metadata-console

pli888 committed:
Fix functional tests in ReplaceFileUrlSubstringWithPrefixCest

pli888 committed:
Remove --doi as a command line parameter

We don't need to provide a DOI since all datasets should have their
ftp_site and file locations URL updated. If a specific dataset
should not have these changes then include them in a list of
excluded DOIs.

pli888 committed:
Add conditions to pass external file location URLs

pli888 committed:
Allow stop doi parameter to halt processing

This is to stop DOIs greater than dataset 200001 from having file
locations updated since any others starting in 200 are not real.

pli888 committed:
Tidy up code

pli888 committed:
Fix functional tests

pli888 committed:
Update documentation

pli888 committed:
Fix assertions in functional tests

pli888 committed:
Fix other functional tests

pli888 committed:
Function actionFileSize should use DatasetFilesUpdater class

pli888 committed:
Correct exclude parameter

pli888 committed:
Update documentation for dev testing

pli888 committed:
Update documentation with SQL command for dataset 100396

kencho51 reviewed (changes requested):
Hi @pli888,

**praise:** The file urls can be replaced with the supplied prefix in both my dev, staging and live.
**praise:** It is great to the command is executed in dry run mode by default.
**praise:** The readme is well documented.
**praise:** Great to have unit test working after adding `Db` module to the unit test suite, the gigadb_testdata database can be restored after each test.
**issue(typo):** In readme file, line 343, the test DatasetFilesURLUpdaterTest.php`:^testReplaceFileUrlSubstringWithPrefix$` does not exist.
**issue(typo):** In readme file, line 352, `:^tryStopDoi$` should be removed as it is trying to run all tests. 
**question:** What is the purpose of having `--prefix=` parameter in the command? The actual prefix has been determined by the `$newFTPLocationPrefix` in the function `replaceFileLocationsForDataset` and `$newFTPSitePrefix` in the function `replaceFTPSiteForDataset` already. For example, I tried `% docker-compose run --rm files-metadata-console ./yii update/urls --prefix=https://test.com --separator=/pub/ --next=3 --exclude='100020,100039' --apply`, the command can be executed successfuly and the `ftp_site` and `location` will still be replaced by the prefix hard coded in the `DatasetFilesURLUpdater` component class.
**suggestion:** Do not make the `--prefix=` as a mandatory option or perform checking whether the supplied prefix equals to the expected one.
**issue(non-blocking):** After the url replacement, there is no indication of changes, for example, `update at` or `modified at` in the tables.
**suggestion(non-blocking):** Would it be worthy to bring in `update at` or `modified at` in `dataset` table and `file` table in this PR?
**suggestion(non-blocking):** In `replaceFTPSiteForDataset` function, there is a counter parameter `$success`, should change it to `$processed` as in the `replaceFileLocationsForDataset` for the consistency?
**suggestion(non-blocking):** Update the `CHANGELOG` when PR #1287 has been merged.

rija reviewed (changes requested):
Hi @pli888,

**praise**: the batched, safe migration of URLs works fine locally and on my AWS deployment
**praise**: Great that more models classes are added to `gigadb/app/models`
**praise**: Good use of the factory pattern for instantiating `DatasetFilesURLUpdater`
**praise**: Terrific work on the automated test infrastructure for unit and functional test suite

**issue**: after running the url replacement script,  the `ftp_site` link on the dataset page shows a permission error
**suggestion**: as discussed with Chris H, let's remove the ftp_site link from the dataset page as we cannot link to directories on Wasabi for non-Wasabi users

**issue**: The container images created in `gitlab-config-(staging|live)-build.yml`  and `gitlab-config-(staging|live)-deploy` for the `files-metadata-console` project have the wrong tag (`latest`) applied to them. The consequence is, because the configuration settings are baked into the image, we can't distinguish between `staging` an `live` configuration and that may cause subtle but high impact data incident when in use in production.
I guess we were lucky so far at the moment probably because we use the same data on staging and beta live for now and don't use both environment at the same time, but it's going to bite us soon enough.
**suggestion**: use $GIGAB_ENV as tag

**issue**: there is a constant `BUCKET_DIRECTORIES` that hard-code the `live` environment. This is not correct as testing the script on staging environment may impact the live environment. 
**suggestion**: pull the Gigadb environment from environment variable which can then be added to `gigadb/app/tools/files-metadata-console/config/params.php` by the means of new `params.php.dist` in template directory `gigadb/app/tools/files-metadata-console/config-sources`

**issue(non-blocking**): the `--prefix` command line option is actually not necessary because its value doesn't change often when we use the command line (it's only change when we change cloud providers which should be rare occurrence) **suggestion**: Either create a constant (e.g URL_PREFIX) in the controller class or better create an environment variable (in case it's not as rare occurrence as I think it is - e.g: multi-cloud, backups, ....) which can be added to `gigadb/app/tools/files-metadata-console/config/params.php` by the means of new `params.php.dist` in template directory`gigadb/app/tools/files-metadata-console/config-sources`


**issue (caching, non-blocking)**: When I followed the testing instructions for the staging environment, the first dataset processed (10001) was logged as done, but it the dataset page was only partially updated (the ftp site link had the correct updated link, but the file locations were all still pointing to the old ftp server). 
**suggestion**: In the SOP for this tool, we should probably add a final step to restart the web app to reset the cache to be safe. Also, if you are able to reproduce it, you could create a Github ticket so we can later investigate our caching validation code

**issue(test, non-blocking)**: Prefixing the shell command in the functional tests with `echo yes |` is not necessary because the command doesn't prompt user for input (which are the scenarios that kind of prefix is meant for)
**issue(test, non-blocking)**: I was surprised to see several of the functional tests have assertions for file locations containing reference to  DOI `100020` which I believed is **not** in `gigadb_testdata`. I then noticed that those files are attached to dataset of id 211 (DOI 100003). It's a bit confusing, so It would be nice to ensure all the file locations for dataset 100003 reference its DOI and not some other DOIs.

**Issue (non-blocking)**: The methods `updateDbDatasetTable()` and `updateDbFileTable()` are extra code that are not necessary
**suggestion**: alternatives:
The code 
```
$this->updateDbDatasetTable($newFTPSite, $dataset->id);
```
Can be written as (not sure the exact syntax for getLastError, check Yii docs):
```
$dataset->ftp_site =  $newFTPSite ;
$dataset->save() || throw new Exception($dataset->getLastError()); 
```

And the code
```
$this->updateDbFileTable($newFileLocation, $file->id);
```
Can be written as
```
$file->location = $newFileLocation;
$file->save() || throw new Exception($file->getLastError());
```

pli888 committed:
Fix typos in README.md identified by kencho51

pli888 committed:
Use URL_PREFIX Gitlab variable to create params.php

URL_PREFIX variable will be used to update file location URLs.

pli888 committed:
Use URL_PREFIX param to create file location and FTP site

pli888 committed:
Remove use of prefix parameter

Not needed because it is created using URL_PREFIX parameter now.

pli888 committed:
Remove use of prefix parameter from functional tests

pli888 committed:
Update changelog

pli888 committed:
Remove FTP site link on dataset pages

pli888 committed:
Use GIGAB_ENV to tag container images

pli888 committed:
Fetch DEPLOYMENT_ENV Gitlab variable into params.php

pli888 committed:
Use DEPLOYMENT_ENV Gitlab variable to create Wasabi paths

This allows hard-coded Wasabi URL in BUCKET_DIRECTORIES constant
to be removed from use.

pli888 committed:
Remove echo yes | from shell commands

The shell commands used in the functional tests do not prompt for
user input so echo yes | is not required.

pli888 committed:
Replace updateDbDatasetTable() and updateDbFileTable()

pli888 committed:
Update documentation to restart web application

This is required reset the cache after dataset ftp site and file
location URLs have been updated.

pli888 committed:
Update file URL locations for dataset doi 100003 in gigadb_testdata

pli888 committed:
Display message if there are no datasets left to process

pli888 committed:
Remove unnecessary findOne method to get dataset or file

pli888 committed:
Add GIGADB_ENV tag to image to be pulled on deployment

pli888 committed:
Correct command to update URLs on AWS environments

pli888 committed:
Update documentation

pli888 committed:
Add unreleased section into changelog

pli888 commented:
>issue(typo): In readme file, line 343, the test DatasetFilesURLUpdaterTest.php:^testReplaceFileUrlSubstringWithPrefix$ does not exist.
>issue(typo): In readme file, line 352, :^tryStopDoi$ should be removed as it is trying to run all tests.

These two typos have been corrected in README.md.

>question: What is the purpose of having --prefix= parameter in the command? The actual prefix has been determined by the $newFTPLocationPrefix in the function replaceFileLocationsForDataset and $newFTPSitePrefix in the function replaceFTPSiteForDataset already. For example, I tried % docker-compose run --rm files-metadata-console ./yii update/urls --prefix=https://test.com --separator=/pub/ --next=3 --exclude='100020,100039' --apply, the command can be executed successfuly and the ftp_site and location will still be replaced by the prefix hard coded in the DatasetFilesURLUpdater component class.
suggestion: Do not make the --prefix= as a mandatory option or perform checking whether the supplied prefix equals to the expected one.

The code has been updated so that the --prefix parameter is not required now.

> issue(non-blocking): After the url replacement, there is no indication of changes, for example, update at or modified at in the tables.
suggestion(non-blocking): Would it be worthy to bring in update at or modified at in dataset table and file table in this PR?

As discussed at a recent VOH, this work will be done in a more holistic manner in the future.

>suggestion(non-blocking): In replaceFTPSiteForDataset function, there is a counter parameter $success, should change it to $processed as in the replaceFileLocationsForDataset for the consistency?

I've not made any code changes - the variable names seem clear to me.

>suggestion(non-blocking): Update the CHANGELOG when PR https://github.com/gigascience/gigadb-website/pull/1287 has been merged.

CHANGELOG.md has been updated with an unreleased section. Maybe @rija can update CHANGELOG.md when this PR has been merged.

pli888 commented:
>issue: after running the url replacement script, the ftp_site link on the dataset page shows a permission error
suggestion: as discussed with Chris H, let's remove the ftp_site link from the dataset page as we cannot link to directories on Wasabi for non-Wasabi users

The view class for datasets has been updated so dataset pages do not display FTP site link now.

>issue: The container images created in gitlab-config-(staging|live)-build.yml and gitlab-config-(staging|live)-deploy for the files-metadata-console project have the wrong tag (latest) applied to them. The consequence is, because the configuration settings are baked into the image, we can't distinguish between staging an live configuration and that may cause subtle but high impact data incident when in use in production.
I guess we were lucky so far at the moment probably because we use the same data on staging and beta live for now and don't use both environment at the same time, but it's going to bite us soon enough.
suggestion: use $GIGAB_ENV as tag

The $GIGADB_ENV tag is now used to tag and select container images in the build and deploy yml files. The $GIGADB_ENV tag is also now required in order to identify which container image to use in order to use the files metadata console tool.

>issue: there is a constant BUCKET_DIRECTORIES that hard-code the live environment. This is not correct as testing the script on staging environment may impact the live environment.
suggestion: pull the Gigadb environment from environment variable which can then be added to gigadb/app/tools/files-metadata-console/config/params.php by the means of new params.php.dist in template directory gigadb/app/tools/files-metadata-console/config-sources

Developers will now need to create a Gitlab secret variable called `DEPLOYMENT_ENV=dev` for dev environment in gigadb-website project-level if they have not already done so.
The `DEPLOYMENT_ENV` variable has been added into `params.php.dist` and is used to create Wasabi bucket path for file location and FTP site URLs instead of using hard-coded URL in BUCKET_DIRECTORIES constant.

>issue(non-blocking): the --prefix command line option is actually not necessary because its value doesn't change often when we use the command line (it's only change when we change cloud providers which should be rare occurrence) suggestion: Either create a constant (e.g URL_PREFIX) in the controller class or better create an environment variable (in case it's not as rare occurrence as I think it is - e.g: multi-cloud, backups, ....) which can be added to gigadb/app/tools/files-metadata-console/config/params.php by the means of new params.php.dist in template directorygigadb/app/tools/files-metadata-console/config-sources

A Gitlab secret variable called `URL_PREFIX=https://s3.ap-northeast-1.wasabisys.com/` has been created in Forks and Upstream. For dev testing, you will need to re-run `docker-compose run --rm configure` to update `.secrets` with URL_PREFIX variable. The URL_PREFIX parameter is now used in DatasetFilesURLUpdater component class. Finally, use of `--prefix` as a command line parameter in UpdateController class and in tests have been removed.

>issue (caching, non-blocking): When I followed the testing instructions for the staging environment, the first dataset processed (10001) was logged as done, but it the dataset page was only partially updated (the ftp site link had the correct updated link, but the file locations were all still pointing to the old ftp server).
suggestion: In the SOP for this tool, we should probably add a final step to restart the web app to reset the cache to be safe. Also, if you are able to reproduce it, you could create a Github ticket so we can later investigate our caching validation code

I confirm that if you visit http://gigadb-staging.gigatools.net/dataset/100001 on your web browser before and after updating its file location URLs using the files metadata console tool, you will see that its file location links always begin with https://ftp.cngb.org/. However, if you browse http://gigadb-staging.gigatools.net/dataset/100002 only after updating its file location URLs, you will see that dataset 100002 page contains file links to https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets. This suggests that there is an issue with the caching validation code. A Github ticket has been created - see https://github.com/gigascience/gigadb-website/issues/1335.

README.md updated with documentation to restart web application after using this tool when on AWS environment.

>issue(test, non-blocking): Prefixing the shell command in the functional tests with echo yes | is not necessary because the command doesn't prompt user for input (which are the scenarios that kind of prefix is meant for)

echo yes | has been removed from shell commands in functional tests.

>issue(test, non-blocking): I was surprised to see several of the functional tests have assertions for file locations containing reference to DOI 100020 which I believed is not in gigadb_testdata. I then noticed that those files are attached to dataset of id 211 (DOI 100003). It's a bit confusing, so It would be nice to ensure all the file locations for dataset 100003 reference its DOI and not some other DOIs.

File locations URLs for dataset 100003 in gigadb_testdata have been updated with more intuitive links.

>Issue (non-blocking): The methods updateDbDatasetTable() and updateDbFileTable() are extra code that are not necessary
suggestion: alternatives:

The changes have been implemented but there is no getLastError() method for active record objects. Instead, I've opted to return all attribute errors as a string which will be useful information.

rija reviewed (commented):
Hi @pli888,

**praise**: the code change looks fine
**praise**: the tool works as expected on my local deployment
**issue**: on staging, when I run 
```
docker run --rm "registry.gitlab.com/gigascience/forks/rija-gigadb-website/production-files-metadata-console:staging" ./yii update/urls --separator=/pub/ --exclude='100375,100396,100446,100584,100747,100957,102311' --stop=200002 --next=30 --apply
```
 I got the following error: 
```
	Transforming file locations for dataset 100027...
yii\base\ErrorException: Array to string conversion in /gigadb/app/tools/files-metadata-console/components/DatasetFilesURLUpdater.php:118
Stack trace:
#0 [internal function]: yii\base\ErrorHandler->handleError(2, 'Array to string...', '/gigadb/app/too...', 118)
#1 /gigadb/app/tools/files-metadata-console/components/DatasetFilesURLUpdater.php(118): implode(Array)
#2 /gigadb/app/tools/files-metadata-console/controllers/UpdateController.php(121): app\components\DatasetFilesURLUpdater->replaceFileLocationsForDataset('100027', '/pub/')
#3 [internal function]: app\controllers\UpdateController->actionUrls()
#4 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#5 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#6 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Controller.php(180): yii\base\Controller->runAction('urls', Array)
#7 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Module.php(552): yii\console\Controller->runAction('urls', Array)
#8 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('update/urls', Array)
#9 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction('update/urls', Array)
#10 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Application.php(384): yii\console\Application->handleRequest(Object(yii\console\Request))
#11 /gigadb/app/tools/files-metadata-console/yii(20): yii\base\Application->run()
#12 {main}
** Rolling back transaction
PHP Warning 'yii\base\ErrorException' with message 'Array to string conversion'

in /gigadb/app/tools/files-metadata-console/components/DatasetFilesURLUpdater.php:118

Stack trace:
#0 [internal function]: yii\base\ErrorHandler->handleError(2, 'Array to string...', '/gigadb/app/too...', 118)
#1 /gigadb/app/tools/files-metadata-console/components/DatasetFilesURLUpdater.php(118): implode(Array)
#2 /gigadb/app/tools/files-metadata-console/controllers/UpdateController.php(121): app\components\DatasetFilesURLUpdater->replaceFileLocationsForDataset('100027', '/pub/')
#3 [internal function]: app\controllers\UpdateController->actionUrls()
#4 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#5 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#6 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Controller.php(180): yii\base\Controller->runAction('urls', Array)
#7 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Module.php(552): yii\console\Controller->runAction('urls', Array)
#8 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('update/urls', Array)
#9 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction('update/urls', Array)
#10 /gigadb/app/tools/files-metadata-console/vendor/yiisoft/yii2/base/Application.php(384): yii\console\Application->handleRequest(Object(yii\console\Request))
#11 /gigadb/app/tools/files-metadata-console/yii(20): yii\base\Application->run()
#12 {main}
[centos@ip-10-99-0-14 ~]$ 
```
**praise**: When I add the problematic DOI above to the exclude list, the tool is working fine and after triggering `sd_gigadb` on the gitlab pipeline, the website reflects the expected changes on staging.
**praise**: good idea adding an "Unreleased" section to the CHANGELOG. I think we should keep adding our PR headline in there for now until we've figured out a proper release management

kencho51 commented:
Hi @pli888,

**praise:** The `--prefix` option has been removed.
**issue:** When I tried to spin up `files-metadata-console` tool, I got this error:
```
% ./up.sh
% cd gigadb/app/tools/files-metadata-console
% ./configure
% docker-compose run --rm composer composer install
WARN[0000] a network with name deployment_db-tier exists but was not created for project "files-metadata-console".
Set `external: true` to use an existing network 
network deployment_db-tier was found but has incorrect label com.docker.compose.network set to "db-tier"
```

**suggestion:** It seems the `docker-compose.yml` is trying to create a network inside the `files-metadata-console` application, which is not the intentional usage. This can be fixed by adding `external` parameter to the docker-compose file as below:
```
networks:
  proxy-db-tier:  # represents db-tier network in ops/docker-compose.yml
    name: deployment_db-tier
    external: true
 ```

Therefore, as a result, the would use the existing external network `deployment_db-tier`, instead of creating a new one.

Details of the `external` parameters can be found at https://docs.docker.com/compose/networking/#use-a-pre-existing-network
**issue:** The functional test` tests/functional/ReplaceFileUrlSubstringWithPrefixCest.php` is failing on my dev and staging environment, the failing tests are:
```
Functional Tests (4) --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
✖ ReplaceFileUrlSubstringWithPrefixCest: Try excluding dois from file url changes 3x[F] (0.20s)
✖ ReplaceFileUrlSubstringWithPrefixCest: Try replace file url substring with prefix 5x[F] (0.33s)
✖ ReplaceFileUrlSubstringWithPrefixCest: Try stop doi 2x[F] (0.17s)
.
.
.
1) ReplaceFileUrlSubstringWithPrefixCest: Try excluding dois from file url changes
 Test  tests/functional/ReplaceFileUrlSubstringWithPrefixCest.php:tryExcludingDoisFromFileUrlChanges
 Step  Can see in shell output " Transforming ftp_site for dataset 100002...\nDONE"
 Fail  Failed asserting that '  Transforming ftp_site for dataset 100002...\n
Apply is true - saving dataset ftp_site attributeDONE\n
\n
        Transforming file locations for dataset 100002...\n
DONE (7/7)\n
\n
        Transforming ftp_site for dataset 100005...\n
Apply is true - saving dataset ftp_site attributeDONE\n
\n
        Transforming file locations for dataset 100005...\n
WARNING (1/4)\n
\n
        Transforming ftp_site for dataset 100039...\n
Apply is true - saving dataset ftp_site attributeDONE\n
\n
        Transforming file locations for dataset 100039...\n
DONE (24/24)\n
' contains "    Transforming ftp_site for dataset 100002...
DONE".
```
It seems codeception could not interpret the `\n` as the newline character.

pli888 committed:
Increase location validation length to 1000 chars

This corresponds to varchar 1000 data type for location column
in file table.

pli888 committed:
Improve Exception message

pli888 committed:
Ignore File links containing ensemblgenomes.org

Files containing this host domain do not need to be updated by
file metadata tool.

pli888 committed:
Update validation rules

Validation rules now correspond to data types in File table columns.

pli888 committed:
Fix bug when stop DOI is reached

The tool needs to stop URL processing when the current DOI is equal
to or greater than the stop DOI.

pli888 committed:
Fix bugs when processing live data

More file URL domains have been added for tool to ignore. FTP site
exception message has been improved too.

pli888 commented:
@rija

>I got the following error:
```
Transforming file locations for dataset 100027...
yii\base\ErrorException: Array to string conversion in /gigadb/app/tools/files-metadata-console/components/DatasetFilesURLUpdater.php:118
```

This error has been fixed by updating the location validation rule in the File model class.

pli888 committed:
Join deployment_db-tier network using external option

This option is now required to join a pre-existing network.

pli888 committed:
Fix bug when no stop DOI is provided on cmd-line

File URL transformation is stopped when the stop DOI is not equal
to 0 and when the current DOI is greater than or equal to the stop
DOI.

pli888 committed:
Update documentation

pli888 committed:
Provide better feedback for problematic File location URLs

This is to help updating links in dataset 100027 which has a
problem on staging env.

pli888 committed:
Uncomment external option

Need to upgrade docker compose version in CICD pipeline before
this option can be used.

pli888 committed:
Update try block

pli888 commented:
>docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/functional/ReplaceFileUrlSubstringWithPrefixCest.php

Functional tests are passing again now. If you are running Docker Desktop version 4.22.0 on your Mac then you will probably have to uncomment the `external: true` option in the files-metadata-console docker-compose.yml file.

As discussed in today's VOH, the work to fix the external option error below will be done in a later ticket to expedite merging of current PRs:
```
$ docker-compose run --rm files-metadata-console ./vendor/codeception/codeception/codecept run --debug tests/functional/ReplaceFileUrlSubstringWithPrefixCest.php
WARN[0000] a network with name deployment_db-tier exists but was not created for project "files-metadata-console".
Set `external: true` to use an existing network 
network deployment_db-tier was found but has incorrect label com.docker.compose.network set to "db-tier"
```

As it stands, the File location URL update tool in files-metadata-console works on production data using this command:
```
# 100310 has too many files (19,981) and causes memory error
# 100396 has too many files (197,214) and causes memory error
# 100608 has too many files (12,864) and causes memory error
# 100707 has too many files (16,675) and causes memory error
$ docker run --rm "registry.gitlab.com/$GITLAB_PROJECT/production-files-metadata-console:$GIGADB_ENV" ./yii update/urls --separator=/pub/ --stop=200002 --next=25 --exclude='100310,100396,100608,100707' --apply
```

The excluded datasets will therefore require manual SQL commands to update dataset FTP site and File location links.

rija reviewed (approved):
Hi @pli888,
I can see that all the requested changes have been implemented.
All the tests are passing for me locally and on CI
I'm happy to approve it.
